### ISSUE TITLE
[ENH] Thoughts on creating a "find my k" sub-module within sktime?

### ISSUE THREAD

**Issue by miraep8** [alan-turing-institute/sktime#2901]
Hey Annotation/Segmentation/Changepoints team!

After a few discussions with you all, it seems like a common problem many of us will need to solve to some degree is "finding k" ie the number of change points segments etc in our data.

One idea I had was that rather than finding k within each of our estimators, it might make sense to separate this logic into a different submodule within `sktime.annotation` that people could import and use in addition to each of our estimators.  

Of the top of my head there are a few approaches to finding k that we might want to consider or support (or perhaps we want to just focus on one?):

- extracting `k` through independent inspection of the data.
       - so this might look like a pre-processing step where we find `k` and pass it to our estimator
- extracting `k` by iteratively fitting each of our estimators for a range of `k` and then returning the best one?
       - this might look more like passing our estimator and data to some sort of `tuning` helper?  Might be able to make use of some of the `GridSearch` functionality that already exists in sktime? (though we might need to configure it so it makes sense for evaluating how well an annotation estimator is performing?) 
- *potentially* we could also consider supporting some probability distribution over the number of segments returned (for example, could be modeled as a dirichlet process or something? 
        - this might also look like a pre-preprocessing step, but then we could imagine calculating an expectation value using our estimators instead of a MLE estimate?
        
Anyway - wanted to loop you all in to the discussion so you could also share any ideas or concerns you might have about this? 

@KatieBuc @fkiraly @lmmentel @NoaBenAmi @lielleravid @Lovkush-A  (please `@` anyone I may have forgotten who would also be interested)

**Comment by fkiraly** [comment]
Hm, option 2 would indeed be a kind of `GridSearch` imo, applied to a parameter `k`, subject to a metric for changepoint, segmentation or the like. This is probably easier in the supervised case.

Option 1, I feel, would require some sort of logic that allows an estimator to produce a *parameter* rather than other predictions. I.e., an estimator that has only `fit`, and `get_fitted_params`? And then a kind of pipeline `FittedParamSetter`, with code like this?

```python
kfitter = KFitterSpecial(params)
segmenter = MySegmenter(more_params)

segmenter_with_ktuning = FittedParamSetter("k", kfitter, segmenter)

# or via dunder :-)
segmenter_with_ktuning = kfitter * segmenter
# although how do we pass the information that the parameter passed is k?
```

**Comment by fkiraly** [comment]
More common cases for the `FittedParamSetter` might be great to have in forecasting:

* determining seasonality parameters by a seasonality test, then stick them in a forecaster
* determining `pdq` for an ARIMA model, then stick them in the ARIMA model
* determining stationarity, and conditionally applying differenting or other stationarity making, that would be `FittedParamSetter` with multiplexing or `OptionalPassthrough`, so you would insert a "left" parameter value into a "right" parameter, with different names

**Comment by miraep8** [comment]
Could be addressed by PR #3041

**Comment by lmmentel** [comment]
Cool idea! I somehow cannot resist thinking about this being similar to choosing number of clusters for clustering approaches that require it. Maybe there is some inspiration from sklearn https://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_silhouette_analysis.html# ? I think this would fall into what you are describing in second bullet point.

**Comment by fkiraly** [comment]
@lmmentel, yes, silhouette analysis is a similar example for the "find k" use case.

But, `sklearn` does not formalize as estimators the process that selects a parameter - #3041 is doing that, and I think there´s nothing comparable in `sklearn`.

**Comment by miraep8** [comment]
Personally I think this is related to the design question which came up in the annotation data generation discussion as well of whether or not everything needs to be an estimator.  Obviously there can be advantages to doing so (may help with composition for example), but we could also imagine writing our estimators so they accept functions as an argument (like sklearn's silhouette analysis for example) as well, which could also has its advantages in that we can use some of the great functionality in sklearn and other libraries (or leave it up to the user to define their own function!) while providing a sensible default! 

I am not necessarily coming down on either side strongly, and I am sure there are more advantages to the estimator route that I have not fully considered either!  Just wanted to point out that the question of this particular design choice has come up in a few places now!  Since it seems relevant to annotation stuff I wonder if we should discuss as a group at some point?

**Comment by fkiraly** [comment]
design-wise, it could be like `KNeighborsTimeSeriesRegressor`, which accepts strings, callables, or classes (that are callable). If a class, it works with `sklearn` `get_params` and `set_params`, and parameters can be tuned.

So, there is not necessarily "sides" or mutual exclusion between callables and classes/estimators.
